build.assets/robotest/Makefile

```diff
@@ -0,0 +1,326 @@
+
+$(ROBOTEST_BUILDDIR):
+	mkdir -p $(ROBOTEST_BUILDDIR)
+
+$(ROBOTEST_UPGRADE_IMAGE_CACHE):
+	mkdir -p $(ROBOTEST_UPGRADE_IMAGE_CACHE)
+
+$(ROBOTEST_IMAGEDIR):
+	mkdir -p $(ROBOTEST_IMAGEDIR)
+
+$(ROBOTEST_TELE_CACHE):
+	mkdir -p $(ROBOTEST_TELE_CACHE)
+
+$(ROBOTEST_BUILD_TMP):
+	mkdir -p $(ROBOTEST_BUILD_TMP)
+
+$(ROBOTEST_CACHE_TMP):
+	mkdir -p $(ROBOTEST_CACHE_TMP)
+
+.PHONY:docker-image
+docker-image: ## Create robotest docker image with the current user/group baked in.
+docker-image: $(ROBOTEST_DOCKER_IMAGE_IIDFILE)
+
+$(ROBOTEST_DOCKER_IMAGE_IIDFILE): Dockerfile Makefile
+	docker build \
+		--build-arg BASE=$(ROBOTEST_DOCKER_UPSTREAM_IMAGE) \
+		--build-arg UID=$(shell id -u) \
+		--build-arg GID=$(shell id -g) \
+		--iidfile $(ROBOTEST_DOCKER_IMAGE_IIDFILE) \
+		--tag $(ROBOTEST_DOCKER_IMAGE) .
+
+.PHONY: images
+images: ## Create all images necessary to run robotest CI tests.
+images: image-robotest images-upgrade image-opscenter
+
+.PHONY: image-robotest
+image-robotest: ## Create robotest cluster image using the current tele.
+image-robotest: $(ROBOTEST_IMAGE)
+
+$(ROBOTEST_IMAGE): $(TELE_OUT) $(ROBOTEST_IMAGE_SRC) | $(ROBOTEST_IMAGEDIR) $(ROBOTEST_BUILD_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_BUILD_TMP)))
+	$(eval TGT := $(TMP)/robotest-$(GRAVITY_VERSION).tar)
+	$(TELE_OUT) build \
+		$(ROBOTEST_APP_YAML) \
+		--state-dir=$(PACKAGES_DIR) \
+		--version $(GRAVITY_VERSION) \
+		--output=$(TGT)
+	mv $(TGT) $(ROBOTEST_IMAGE)
+	rm -rf $(TMP)
+
+.PHONY: images-upgrade
+images-upgrade: ## Create/fetch robotest cluster images needed for upgrade testing.
+images-upgrade:
+	$(MAKE) -j $(ROBOTEST_UPGRADE_IMAGES)
+
+.PHONY: image-opscenter
+image-opscenter: ## Move opscenter.tar cluster image into place for test execution.
+image-opscenter: $(ROBOTEST_OPSCENTER_IMAGE)
+
+$(ROBOTEST_OPSCENTER_IMAGE): $(OPSCENTER_OUT) | $(ROBOTEST_IMAGEDIR)
+	cp $(OPSCENTER_OUT) $(ROBOTEST_OPSCENTER_IMAGE)
+
+# The following rule populates the build specific docker image mount
+# from cached images. Fulfills the $(ROBOTEST_UPGRADE_IMAGES) targets.
+# .PRECIOUS because these need to be present through the run target.
+.PRECIOUS: $(ROBOTEST_IMAGEDIR)/robotest-%.tar
+$(ROBOTEST_IMAGEDIR)/robotest-%.tar: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar | $(ROBOTEST_IMAGEDIR) $(ROBOTEST_BUILD_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_BUILD_TMP)))
+	$(eval ATOMIC := $(TMP)/robotest-$*.tar)
+	cp $< $(ATOMIC)
+	mv $(ATOMIC) $@
+	rm -rf $(TMP)
+
+# The following target is a pattern rule for populating the upgrade image cache
+# Fulfills the $(ROBOTEST_UPGRADE_CACHE_IMAGES) targets.
+# .PRECIOUS because it is a cache shared across versions.
+.PRECIOUS: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar
+$(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar: $(ROBOTEST_TELE_CACHE)/tele-% $(ROBOTEST_UPGRADE_BASE_IMAGE_SRC) | $(ROBOTEST_UPGRADE_IMAGE_CACHE) $(ROBOTEST_CACHE_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
```

Comment:
I wonder if we should consider trapping `rm -rf $(TMP)` to make sure that the directory is removed even if, say, tele build below fails. Same for other targets that create temp. dirs.

Reply:
Makefiles have awful signal handling.  We'd need to shell out to bash.  This is possible, but it is adding yet another layer of complexity.  I'd considered this, but was hoping that we could wait until mage/go builds, where signal handling and defer/recover is first class.  Especially with the newly enlarged jenkins workspace giving us buffer for heavy PRs

As of this PR both `$(ROBOTEST_CACHE_TMP)` and `$(ROBOTEST_BUILD_TMP)` are underneath the build root.  The caches, along with any straggling temp files will be nuked by a `make clean`. If the user puts `$(ROBOTEST_CACHE_ROOT)` outside the build dir, leaking tmp directories is a concern -- but that isn't the case for jenkins (yet).  It is more of a concern for local development.  I'd like to defer this until either:

* We move `$(ROBOTEST_CACHE_TMP)` outside `$(BUILDDIR)` or
* We migrate to mage/go for our build files

Reply:
I've refactored out all the tmp stuff into shell scripts with traps at of c91653e. Although I think ithe code is a bit more difficult to undestand, it aligned nicely with containerizing the `tele build` which was needed to handle `~/.tsh` pollution across a large range of `tele`s.